Added support for Leon processor to TFL micro (includes support for big endian systems)

PeteBlackerThe3rd opened the pull request:
I've been working on implementing ML models on the Leon 3 processor which is a space qualified radiation hardened processor. This pull request includes everything necessary for TFL micro to support this new platform.

There are new 3rd party downloads, built target and test target scripts so that projects can be built for the new 'leon' target. This includes the installation and use of TSim, an emulator of the Leon processor family.

The Leon processor family is based upon the older SparcV8 architecture which is big endian. This has required a few minor updates to the TFL micro code so that the correct values are read from the little endian flat buffers used to store TFL models.

Finally the reading of a non-aligned 64 bit int which is not supported on this architectures needed to be converted to a memcpy to avoid crashing on this target.

The included micro_speech example is now building and executing and all tests are passing.

I look forward to any feedback and getting this update included in the main branch.

Commit:
Added built and test support for the Leon 3 processor to the TensorFlow lite micro framework

Commit:
Added Leon compiler and emulator to third party downloads.

Commit:
Added support for big endian systems to TFL micro: weights are converted to big endian on startup, several accesses are made endian safe and non-aligned safe.

Commit:
Cleaned up code now passing clang-tidy checks

Commit:
Cleaned up a few comments and removed redundant code

Commit:
Sorted out whitespace

PeteBlackerThe3rd commented:
Added my surrey.ac.uk email address to my google account. Don't know if this is going to update the CLA but It's worth a shot

petewarden commented on tensorflow/lite/experimental/micro/simple_tensor_allocator.cc:
```diff
@@ -135,10 +135,13 @@ TfLiteStatus SimpleTensorAllocator::AllocateTensor(
       (src_quantization->zero_point()->size() > 0)) {
     result->params.scale = src_quantization->scale()->Get(0);
-    result->params.zero_point = src_quantization->zero_point()->Get(0);
+    memcpy(&result->params.zero_point,
```
We're trying to avoid memcpy where possible, since on some toolchains it requires libc linker support. Could you replace it with either an explicit loop or a small utility function? Thanks!

PeteBlackerThe3rd replied:
I've changed this to a loop, to remove any library dependencies.

petewarden commented on tensorflow/lite/experimental/micro/micro_interpreter.cc:
```diff
@@ -96,6 +96,51 @@ MicroInterpreter::MicroInterpreter(const Model* model,
+
+  // If the system is big endian then convert weights from the flatbuffer from
+  // little to big endian on startup so that it does not need to be done during
+  // inference.
```
Can you add a comment clarifying that this assumes the weight parameters are held in memory that can be modified, rather than being truly read-only?

PeteBlackerThe3rd replied:
Added an additional note here, that the memory must be able to be modified.

petewarden reviewed (changes requested):
This look great, thanks for the contribution! There are a couple of minor comments, otherwise it's good to go.

Commit:
Updates to PR #30362 to address PeteWardens review

Commit:
Fixed automatic downloading and install scripts for getting leon tool chain.

PeteBlackerThe3rd commented:
I've addressed the two comment from your review now. I also found a bug in the code which automatically downloaded the leon toolchain, this didn't show up on my dev system because the tools were already installed. I've modified the download_and_extract.sh bash script so it now supports tar.xz files and will also display an error if an unsupported archive type is downloaded, as opposed to silently failing.

If these latest updates are okay then this is good to merge as far as I can tell.

PeteBlackerThe3rd committed:
Merge branch 'master' into master

petewarden commented on tensorflow/lite/experimental/micro/micro_interpreter.cc:
```diff
@@ -83,9 +83,67 @@ MicroInterpreter::MicroInterpreter(const Model* model,
   context_.ReportError = ReportOpError;
   context_.recommended_num_threads = 1;
+
+  initialization_status_ = AllocateInputAndActTensors();
```
I don't see the implementations of `AllocateInputAndActTensors()` or `AllocateTemporaryTensors`. Are they meant to be part of this PR, or has a future change leaked in here?

PeteBlackerThe3rd replied:
They were originally in this PR but they appear to have been removed in the trunk since I forked this branch. The latest version of micro_interpreter.cc on Tensorflow master doesn't have these methods at all and has other changes too. 

If I remove the calls within the constructor from this PR, then this should be cleared up when resolving the merge conflicts. The latest version calls AllocateTensors() within the Invoke method instead.

Commit:
Updated MicroInterpreter constructor so it will merge more tidily with the latest version

Commit:
Merge branch 'master' of https://github.com/PeteBlackerThe3rd/tensorflow

PeteBlackerThe3rd reviewed (commented):
Agreed we'll resolve this in the merge.

Commit:
Merge branch 'master' of https://github.com/tensorflow/tensorflow

Merging LEON3 support update for TensorFlow lite micro with the latest uptream version

petewarden reviewed (dismissed):
Thanks for the updates, approved!

rthadur commented:
@PeteBlackerThe3rd can you please check build failures ?

PeteBlackerThe3rd commented:
> @PeteBlackerThe3rd can you please check build failures ?

On it. Looks like I added a dependency into tensorflow/lite/kernels/kernel_utils.h source file without updating the bazel build file. My mistake, I only tested the TF lite micro builds not the full TF build.

Currently building a (hopefully) fixed version now.

Commit:
Added missing @flatbuffers dependency to kernel_util library in BUILD file

PeteBlackerThe3rd commented:
I've added the missing dependency now and successfully built a local linux CPU version from this fork. Should be good to go this time.
